### PR TITLE
Add rosparam for specifying the max acceleration

### PR DESCRIPTION
**Pull request by mic-e**
The parameter is in m/s^2.

When the param does not exist, the default of 1000000 ticks/s^2 is used.

### CODE REVIEW

**Comment by padhupradheep** [comment]
Hey ! Thanks for the PR. 
We do not want to give acceleration limit out as a free parameter, this might endanger the hardware if someone mistakenly change it beyond the limits. Thus this feature might not allow us to merge this feature branch to the master anytime in future.

**Comment by mic-e** [comment]
Ah, I see; thanks for the info!

The reason why I added this parameter is because in our case the system becomes very unstable with the high acceleration of 1000000 ticks/s^2 so I decreased the acceleration to 1.0 m/s^2 (530516 ticks/s^2).

Would you consider merging the PR if I enforce that the parameter cannot a limit of max. 1000000 ticks/s^2?

**Comment by padhupradheep** [comment]
Yes we will accept it, with the limit and as well as if you change the unit to `ticks/s^2` !

File: src/neo_omnidrive_socketcan.cpp [hunk 1]
```diff
@@ -151,6 +151,13 @@ class NeoSocketCanNode {
 		m_node_handle.param("measure_torque", m_measure_torque, false);
 		m_node_handle.param("homeing_button", m_homeing_button, 0);
 		m_node_handle.param("diagnostics_prefix", m_diagnostics_prefix, std::string(""));
+		// these values are used to calculate and store max_accel_enc_s for each motor
+		double wheel_radius;
+		if(!m_node_handle.getParam("wheel_radius", wheel_radius)) {
+			throw std::logic_error("missing wheel_radius param");
+		}
+		double max_drive_acc;
+		m_node_handle.param("max_drive_acc", max_drive_acc, 0.0);  // in m/s^2
```
Comment (padhupradheep) [inline src/neo_omnidrive_socketcan.cpp:160]: I believe you wanted to say `// in ticks/s^2` 

**Edit:** On a brief look on the other parts of the code, I'm wondering whether this parameter is in `ticks/s^2`

**Review by padhupradheep — CHANGES_REQUESTED** [review]

**Comment by madMAx43v3r** [comment]
> Ah, I see; thanks for the info!
> 
> The reason why I added this parameter is because in our case the system becomes very unstable with the high acceleration of 1000000 ticks/s^2 so I decreased the acceleration to 1.0 m/s^2 (530516 ticks/s^2).

Usually there are acceleration limits further up the chain, in the planner for example. That's why we don't use the motor parameters to limit acceleration of the platform, they are just a safety limit for us.

Is it not possible to add a limit in your planner?

**Comment by mic-e** [comment]
> Is it not possible to add a limit in your planner?

Our local planner does have its own acceleration limits, but these will not take effect when

- the local planner's command is cut down due to (soft) safety constraint violations
- the local planner is not used during keyboard/joystick teleopping

thus the additional acceleration limit closer to the hardware.

File: src/neo_omnidrive_socketcan.cpp (hunk 1 above)
Comment (mic-e) [reply src/neo_omnidrive_socketcan.cpp:160]: The value is indeed in m/s^2; I thought that a more human-readable version in SI units would be preferable.

Down in line 216 you can see the conversion from m/s^2 to ticks/s^2 - this is also the reason why I need to read the radius parameter.

I have prepared an alternative version of this PR where instead the acceleration is specified in ticks/s^2 for each drive motor individually; your choice :)

**Comment by mic-e** [comment]
Hi, it looks like this PR has stalled out. Is there anything that still needs doing?

Do you want the PR like this (acceleration specified globally in m/s^2), or do you want it with the acceleration specified for each wheel individually in ticks/s^2?

**Comment by madMAx43v3r** [comment]
Yes, `ticks/s^2`, because it's just a safety limit. Setting a max. acceleration in `m/s^2` gives the impression that it's kinematically correct, but it isn't. If that limit is reached it will invalidate the kinematics, ie. one of the wheels will be dragged over the ground. This should never happen in normal operation.

**Comment by madMAx43v3r** [comment]
> Our local planner does have its own acceleration limits, but these will not take effect when
> 
>     * the local planner's command is cut down due to (soft) safety constraint violations
> 
>     * the local planner is not used during keyboard/joystick teleopping
> 
> 

We have an additional low pass filter in the teleop node for that. In general any acceleration limits should be applied on x/y + rot velocity, not on individual wheel velocities.

**Comment by mic-e** [comment]
I see; I understand your point and I think it makes no sense to have this feature in the upstream.

I now think that the correct place for enforcing the acceleration limit would be in the `neo_omnidrive_node`.

Thanks for your time anyways!

**Pull request closed without merging**